refactor service_manager.go

mikhail-aws opened the pull request:
Note:
- major change in service_manager.go and service_manager_test.go, completely rewritten, diff might be hard to read. Adding links to source files for readers [service_manager](https://github.com/mikhail-aws/aws-application-networking-k8s/blob/bug-fix-317/pkg/deploy/lattice/service_manager.go) and [test](https://github.com/mikhail-aws/aws-application-networking-k8s/blob/bug-fix-317/pkg/deploy/lattice/service_manager_test.go)

Testing:
- unit tests
- e2e tests
```
Ran 13 of 13 Specs in 2343.684 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (2344.70s)
```

mikhail-aws commented:
#342

mikhail-aws commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,285 @@ package lattice
+type GetTagsReq = vpclattice.ListTagsForResourceInput
+type GetTagsResp = vpclattice.ListTagsForResourceOutput
+type ListAssocsReq = vpclattice.ListServiceNetworkServiceAssociationsInput
+type SnSvcAssocSummary = vpclattice.ServiceNetworkServiceAssociationSummary
```
only for these two? tags and assocs?

solmonk replied:
Other changes are acceptable I think. Just want to make sure we are not obfuscating API names

erikfuller commented on docs/developer.md:
```diff
@@ -1,57 +1,113 @@
-## Developer Guide
+# Developer Guide
```
I'm a little confused - why is this file showing up in the diff? I'm ont seeing any changes from what's on main?

erikfuller commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,301 @@ package lattice
-			}
+// update SN-Svc associations, if svc has no SN associations will delete all of them
+// does not delete associations that are not tagged by controller
+func (m *defaultServiceManager) updateSnSvcAssocs(ctx context.Context, svc *Service, svcSum *SvcSummary) error {
```
I'm not sure we should be using this method for both ```Create/Update``` and ```Delete```. On delete, shouldn't we be deleting all associations whether or not we created them?

mikhail-aws replied:
Actually I'm using this on Update and Delete but not create. But I agree that delete should be more straightforward with deletion. So create/update/delete will have own little method.

erikfuller commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,301 @@ package lattice
+		ServiceIdentifier: svcSum.Id,
 	}
-	return nil
+	_, err = m.cloud.Lattice().DeleteServiceWithContext(ctx, &delInput)
```
At this point, could we be dealing with a service that doesn't have our management tags? Should we be checking? How do you envision upgrades working from a previous version of the controller to this one?

mikhail-aws replied:
There is not tag check on service deletion, and it wasnt before as far as I can tell. About moving forward - while we still at alpha stage and breaking changes happens, there is little concern about backward compatibility. Starting from Beta we cant do that.

mikhail-aws replied:
My initial thought was to make something like "ensureTag" that will go through existing associations and tag them if not tagged. But then controller does not about any associations right now. If I restart controller local storage is gone, and next reconcile event can already mutate list of SN's. So at this point we need to have any tag, that can be a starting point for future upgrades.

mikhail-aws replied:
will do tags in another PR, removed from this

erikfuller reviewed (changes requested):
Thanks for putting this together. I'm still working my way through the tests but wanted to get my initial comments up. I think the big question is whether or not the proposed tag actually improves the situation - so far I'm not convinced. I do appreciate the refactoring changes in ```service_manager.go```, it's much more readable and has less duplication.

mikhail-aws commented on pkg/model/lattice/servicenetwork.go:
```diff
@@ -5,6 +5,7 @@ import (
 
 const (
+	TagManagedBy                = "ManagedBy"
 	K8SServiceNetworkOwnedByVPC = "K8SServiceNetworkOwnedByVPC"
 	K8SServiceOwnedByVPC        = "K8SServiceOwnedByVPC"
```
yes, it's a larger refactoring, need follow up on this

mikhail-aws commented on pkg/aws/cloud.go:
```diff
@@ -37,37 +68,89 @@ func NewCloud(log gwlog.Logger) (Cloud, error) {
 
-var _ Cloud = &defaultCloud{}
+// Used in testing and mocks
+func NewDefaultCloud(l services.Lattice, e services.EKS, cfg CloudConfig) Cloud {
```
agree, a bit too much

mikhail-aws commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,285 @@ package lattice
 
 import (
```
will do 2 prs

mikhail-aws commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,301 @@ package lattice
+	}
 
-		}
+	svc.Spec.ServiceNetworkNames = []string{}
```
I was having same debate with myself. Since you also mentioning this, I will change it

mikhail-aws replied:
split update and delete into different methods, no longer mutate spec

mikhail-aws commented:
@erikfuller @solmonk 

- updated this PR with only refactoring for service_manager, removed bug fix
- re-ran e2e tests with latest changed from main

erikfuller commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,322 @@ package lattice
-		}
-	} else {
-		return nil, err
+func svcStatusFromCreateSvcResp(resp *CreateSvcResp) ServiceStatus {
```
I see somewhere else similar code is used, I would suggest changing the signature to take the arn, id, and dns - I don't think for example the ```resp``` value here can be nil and identical logic lives in ```updateServiceAndAsociations```.

erikfuller commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,322 @@ package lattice
+		return status
 	}
-	return nil, nil
+	status.Arn = aws.StringValue(resp.Arn)
```
So...I think we should rename all these to ```service``` and reserve ```status``` for when we're actually working with some kind of status.

mikhail-aws replied:
Status with kind of status make sense :D
It comes from latticemodel, I just created 2 aliases to Service and ServiceStatus. Same as before. A hammer of refactoring will get there one day.

erikfuller commented on pkg/deploy/lattice/service_manager_test.go:
```diff
@@ -3,1035 +3,296 @@ package lattice
+	// This test covers creation/deletion for multiple SN's
+	// Service is associated with sn-keep and sn-delete,
+	// for update we need to delete sn-delete and add sn-add
+	t.Run("update service's associations", func(t *testing.T) {
```
This is a nice test!

erikfuller commented on pkg/deploy/lattice/service_manager_test.go:
```diff
@@ -3,1035 +3,296 @@ package lattice
+
+		// assert that we call create association
+		lat.EXPECT().
+			CreateServiceNetworkServiceAssociationWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
```
Would be nice to include some assertions in this test on arguments

mikhail-aws replied:
Request part is covered in smaller unit test TestCreateSvcReq

erikfuller commented on pkg/deploy/lattice/service_manager_test.go:
```diff
@@ -3,1035 +3,296 @@ package lattice
+
+		// assert we delete association and service
+		lat.EXPECT().
+			DeleteServiceNetworkServiceAssociationWithContext(gomock.Any(), gomock.Any(), gomock.Any()).
```
Worth adding some assertions on args?

erikfuller commented on pkg/deploy/lattice/service_manager_test.go:
```diff
@@ -3,1035 +3,296 @@ package lattice
 
-			}
-			mockVpcLatticeSess.EXPECT().ListServiceNetworkServiceAssociationsAsList(ctx, gomock.Any()).Return(listMeshServiceAssociationsOutput, nil)
+func TestSnSvcAssocsDiff(t *testing.T) {
```
Thanks for adding these - good to have full coverage on this logic.

erikfuller reviewed (changes requested):
Thanks for the PR, very readable! I've left a number of suggestions, but I think the ones I care most about are:

#### 1. Prefer pointer return types
In addition to being able to return ```nil``` instead of having a placeholder empty return, this will help ensure calling code can't act on an empty value, for example if the calling code forgets to check ```err```. With ```nil```, any attempts to act on the value will cause the code to fail, most likely very loudly, which I think is better than taking an incorrect action. Example:

```golang
func (...) updateServiceAndAssociations(...) (ServiceStatus, error) {
```
becomes
```golang
func (...) updateServiceAndAssociations(...) (*ServiceStatus, error) {
```

#### 2. Rename ```status ServiceStatus``` variables to ```service```
These are more services than statuses, and given we have actual status values around, it's confusing to read these variable names as they are right now.

Other comments can be considered optional, but if you feel like making changes based on them go for it.

mikhail-aws commented:
@erikfuller, thank you for review

> 1. Prefer pointer return types

Your reasoning is correct about failing on nil is more obvious than failing on empty struct. But it's debatable that having nil exception is better than empty struct. Also it's not exactly how errors should be handled in go. We should rely on error checking. Error checking can be (and should be) enforced by lint. Example https://github.com/kisielk/errcheck. Part of #321 

What I consider when return pointer or struct:

 - simple, small structs with primitive values always return by value. In go everything is mutable and cheap concurrency tend to create lots of goroutines with shared (mutable!) objects. So passing copies of data around is safer, and preferable. (frowning at lattice-datastore). Mutability one of the major concerns.
 - use pointer for complex structs with internal states, or expensive to copy
 - use pointer when nil is valid option, an analog to Optional<T>. In this case when we create service and there is no error it will return valid response. (aws sdk is full of pointers, everything is optional...)

Returning struct here struct is fine. I can rewrite code a bit to make it less "verbose"

mikhail-aws commented on pkg/aws/cloud.go:
```diff
@@ -37,34 +47,41 @@ func NewCloud(log gwlog.Logger) (Cloud, error) {
+	return c.cfg
 }
 
 func (d *defaultCloud) GetEKSClusterVPC(name string) string {
```
Will rename receiver, but it's used and will be used more in my planned refactoring for config propagation. #310 
```
pkg/deploy/lattice/service_manager.go
84:	sn, err := m.datastore.GetServiceNetworkStatus(snName, m.cloud.Config().AccountId)
```

mikhail-aws replied:
sorry, my bad I thought you were asking about Config, but it was GetEKSClusterVPC(), yes it's not used. It's on my personal todo list

mikhail-aws commented on pkg/deploy/lattice/error.go:
```diff
@@ -1,9 +1,13 @@
 	LATTICE_RETRY              = "LATTICE_RETRY"
 )
+
+var RetryErr = errors.New(LATTICE_RETRY)
```
Works fine for golang standard library, io errors are reused. https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/internal/oserror/errors.go

mikhail-aws commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,322 @@ package lattice
+}
 
+func (m *defaultServiceManager) createAssociation(ctx context.Context, svcId *string, snName string) error {
+	sn, err := m.datastore.GetServiceNetworkStatus(snName, m.cloud.Config().AccountId)
```
I think current implementation of datastore introduces more not-handled edge cases than solve problems. We are scratching a surface here. I believe we need to get back to drawing board at some point about entire flow.

mikhail-aws commented on pkg/deploy/lattice/service_manager.go:
```diff
@@ -2,318 +2,322 @@ package lattice
-	svcName := latticestore.LatticeServiceName(service.Spec.Name, service.Spec.Namespace)
-	serviceSummary, err := s.findServiceByName(ctx, svcName)
+// find service by name, if not found returns nil without error
+func (m *defaultServiceManager) getService(ctx context.Context, svcName string) (*vpclattice.ServiceSummary, error) {
```
In original code we dont store/get service from cache, will keep it in this PR, since it's only code refactoring with same semantics.

mikhail-aws commented:
@erikfuller 
- renamed ServiceStatus alias to ServiceInfo [link](https://github.com/aws/aws-application-networking-k8s/blob/38d27477508eb341e12a5af8a9b8aa18b1d0fead/pkg/deploy/lattice/service_manager.go#L18)
- renamed "status" variables into svcInfo [link](https://github.com/aws/aws-application-networking-k8s/blob/38d27477508eb341e12a5af8a9b8aa18b1d0fead/pkg/deploy/lattice/service_manager.go#L118)
- added arguments assertion to Create and Delete integ tests

erikfuller reviewed (approved):
lgtm
